pkg/transforms/mqttsecret.go

```diff
@@ -0,0 +1,265 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package transforms
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	MQTT "github.com/eclipse/paho.mqtt.golang"
+	"github.com/edgexfoundry/app-functions-sdk-go/appcontext"
+	"github.com/edgexfoundry/app-functions-sdk-go/pkg/util"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients"
+)
+
+// MQTTSecretSender ...
+type MQTTSecretSender struct {
+	client         MQTT.Client
+	mqttConfig     MQTTSecretConfig
+	persistOnError bool
+	opts           *MQTT.ClientOptions
+	lastRetrieved  time.Time
+}
+
+// MQTTSecretConfig ...
+type MQTTSecretConfig struct {
+	// BrokerAddress should be set to the complete broker address i.e. mqtts://mosquitto:8883/mybroker
+	BrokerAddress string
+	// ClientId to connect with the broker with.
+	ClientId string
+	// The name of the path in secret provider to retrieve your secrets
+	SecretPath string
+	// AutoReconnect indicated whether or not to retry connection if disconnected
+	AutoReconnect bool
+	// Topic that you wish to publish to
+	Topic string
+	// QoS for MQTT Connection
+	QoS byte
+	// Retain setting for MQTT Connection
+	Retain bool
+	// SkipCertVerify
+	SkipCertVerify bool
+	// AuthMode indicates what to use when connecting to the broker. Options are "none", "cacert" , "usernamepassword", "clientcert".
+	// If a CA Cert is exists in the SecretPath then it will be used for all modes except "none".
+	AuthMode string
+}
+type mqttSecrets struct {
+	username     string
+	password     string
+	keypemblock  []byte
+	certpemblock []byte
+	capemblock   []byte
+}
+
+const (
+	AuthModeNone             = "none"
+	AuthModeUsernamePassword = "usernamepassword"
+	AuthModeCert             = "clientcert"
+	AuthModeCA               = "cacert"
+	// Name of the keys to look for in secret provider
+	MQTTSecretUsername   = "username"
+	MQTTSecretPassword   = "password"
+	MQTTSecretClientKey  = "clientkey"
+	MQTTSecretClientCert = AuthModeCert
+	MQTTSecretCACert     = AuthModeCA
+)
+
+// NewMQTTSecretSender ...
+func NewMQTTSecretSender(mqttConfig MQTTSecretConfig, persistOnError bool) *MQTTSecretSender {
+	opts := MQTT.NewClientOptions()
+
+	opts.AddBroker(mqttConfig.BrokerAddress)
+	opts.SetClientID(mqttConfig.ClientId)
+	opts.SetAutoReconnect(mqttConfig.AutoReconnect)
+	//avoid casing issues
+	mqttConfig.AuthMode = strings.ToLower(mqttConfig.AuthMode)
+	sender := &MQTTSecretSender{
+		client:         nil,
+		mqttConfig:     mqttConfig,
+		persistOnError: persistOnError,
+		opts:           opts,
+	}
+
+	return sender
+}
+func (sender *MQTTSecretSender) getSecrets(edgexcontext *appcontext.Context) (*mqttSecrets, error) {
+	// No Auth? No Problem!...No secrets required.
+	if sender.mqttConfig.AuthMode == AuthModeNone {
+		return nil, nil
+	}
+
+	secrets, err := edgexcontext.GetSecrets(sender.mqttConfig.SecretPath)
```

Comment:
Is there a reason you're not passing in the keys MQTTSecretUsername, MQTTSecretPassoword, ... so on?  This would give an additional level of validation because it would preemptively fail if the keys weren't set.